src/memo/memo.ml

```diff
@@ -632,8 +635,24 @@ let dep_node (type i o f) (t : (i, o, f) t) inp =
     Store.set t.cache inp dep_node;
     dep_node
 
+module Cache_lookup_result = struct
+  type ('a, 'ivar) t =
+    | New_attempt of Running_state.t * 'ivar
+    | Waiting of Running_state.t * 'ivar
```

Comment:
Looks like you're only ever accessing the `sample` field of `Running_state.t`, so you can just store `Dag.node` here directly.

Reply:
Same in the `New_attempt` constructor.

Reply:
I thought it's distasteful to not return the information the caller was about to need, so I instead made them use it in my last commit. We now pass `running_state` to wherever it's needed.

Reply:
In the `Waiting` case it still was unused so I made the change to store `Dag.node` there.

Reply:
Looks good!